File: VoidMail/Services/DeepgramService.swift
import Foundation
import AVFoundation

// MARK: - Deepgram Text-to-Speech Service
// Converts email text to audio using Deepgram's TTS API.

@MainActor
class DeepgramService: ObservableObject {
    static let shared = DeepgramService()

    private let apiKey: String = {
        guard let key = Bundle.main.object(forInfoDictionaryKey: "DEEPGRAM_API_KEY") as? String, !key.isEmpty else {
            return ""
        }
        return key
    }()
    private let baseURL = "https://api.deepgram.com/v1/speak"

    @Published var isGenerating = false
    @Published var isPlaying = false
    @Published var playbackProgress: Double = 0
    @Published var audioDuration: Double = 0
    @Published var error: String?

    private var audioPlayer: AVAudioPlayer?
    private var progressTimer: Timer?
    private var cachedAudioURLs: [String: URL] = [:]  // Cache audio by email ID

    // MARK: - Generate Audio from Email

    /// Generates TTS audio for an email body. Returns the local file URL.
    /// Splits long text into chunks to avoid Deepgram's payload size limit.
    func generateAudio(emailId: String, text: String, voice: String = "aura-asteria-en") async -> URL? {
        // Return cached audio if already generated
        if let cached = cachedAudioURLs[emailId] {
            return cached
        }

        isGenerating = true
        error = nil

        guard !apiKey.isEmpty else {
            error = "Deepgram API key not configured"
            isGenerating = false
            return nil
        }

        do {
            // Trim to reasonable max and split into chunks
            let trimmedText = String(text.prefix(5000))
            let chunks = splitIntoChunks(trimmedText, maxLength: 1500)

            var allAudioData = Data()

            for chunk in chunks {
                guard let chunkData = try await fetchTTSAudio(text: chunk, voice: voice) else {
                    continue
                }
                allAudioData.append(chunkData)
            }

            guard !allAudioData.isEmpty else {
                error = "Failed to generate audio"
                isGenerating = false
                return nil
            }

            // Save combined audio to temp file
            let tempDir = FileManager.default.temporaryDirectory
            let fileURL = tempDir.appendingPathComponent("voidmail_tts_\(emailId).mp3")
            try allAudioData.write(to: fileURL)
            cachedAudioURLs[emailId] = fileURL

            isGenerating = false
            return fileURL

        } catch {
            debugLog("[DeepgramService] generateAudio error: \(error.localizedDescription)")
            self.error = "Failed to generate audio"
            isGenerating = false
            return nil
        }
    }

    /// Fetch TTS audio for a single text chunk from Deepgram
    private func fetchTTSAudio(text: String, voice: String) async throws -> Data? {
        let urlString = "\(baseURL)?model=\(voice)&encoding=mp3"
        guard let url = URL(string: urlString) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("Token \(apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 30

        let body: [String: Any] = ["text": text]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            debugLog("[DeepgramService] Chunk TTS error \(code)")
            return nil
        }

        return data
    }

    /// Split text into chunks at sentence boundaries, respecting maxLength
    private func splitIntoChunks(_ text: String, maxLength: Int) -> [String] {
        guard text.count > maxLength else { return [text] }

        var chunks: [String] = []
        var remaining = text

        while !remaining.isEmpty {
            if remaining.count <= maxLength {
                chunks.append(remaining)
                break
            }

            // Find the last sentence-ending punctuation within maxLength
            let searchRange = remaining.prefix(maxLength)
            var splitIndex = searchRange.startIndex

            // Look for sentence boundaries: . ! ? followed by space or end
            for char in [". ", "! ", "? ", ".\n", "!\n", "?\n"] {
                if let range = searchRange.range(of: char, options: .backwards) {
                    let candidate = range.upperBound
                    if candidate > splitIndex {
                        splitIndex = candidate
                    }
                }
            }

            // If no sentence boundary found, split at last space
            if splitIndex == searchRange.startIndex {
                if let spaceRange = searchRange.range(of: " ", options: .backwards) {
                    splitIndex = spaceRange.upperBound
                } else {
                    // No space found — hard split
                    splitIndex = remaining.index(remaining.startIndex, offsetBy: maxLength)
                }
            }

            let chunk = String(remaining[remaining.startIndex..<splitIndex]).trimmingCharacters(in: .whitespacesAndNewlines)
            if !chunk.isEmpty {
                chunks.append(chunk)
            }
            remaining = String(remaining[splitIndex...]).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        return chunks
    }

    // MARK: - Play Audio

    func play(url: URL) {
        do {
            // Configure audio session
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
            try AVAudioSession.sharedInstance().setActive(true)

            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.prepareToPlay()
            audioDuration = audioPlayer?.duration ?? 0
            audioPlayer?.play()
            isPlaying = true

            // Start progress tracking
            startProgressTracking()
        } catch {
            debugLog("[DeepgramService] play error: \(error.localizedDescription)")
            self.error = "Failed to play audio"
        }
    }

    // MARK: - Pause

    func pause() {
        audioPlayer?.pause()
        isPlaying = false
        stopProgressTracking()
    }

    // MARK: - Resume

    func resume() {
        audioPlayer?.play()
        isPlaying = true
        startProgressTracking()
    }

    // MARK: - Stop

    func stop() {
        audioPlayer?.stop()
        audioPlayer?.currentTime = 0
        isPlaying = false
        playbackProgress = 0
        stopProgressTracking()
    }

    // MARK: - Toggle Play/Pause

    func togglePlayPause(url: URL) {
        if isPlaying {
            pause()
        } else if audioPlayer != nil && audioPlayer?.currentTime ?? 0 > 0 {
            resume()
        } else {
            play(url: url)
        }
    }

    // MARK: - Seek

    func seek(to progress: Double) {
        guard let player = audioPlayer else { return }
        player.currentTime = progress * player.duration
        playbackProgress = progress
    }

    // MARK: - Progress Tracking

    private func startProgressTracking() {
        stopProgressTracking()
        progressTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self, let player = self.audioPlayer else { return }
                if player.isPlaying {
                    self.playbackProgress = player.currentTime / player.duration
                } else if player.currentTime >= player.duration - 0.1 {
                    // Playback finished
                    self.isPlaying = false
                    self.playbackProgress = 0
                    self.stopProgressTracking()
                }
            }
        }
    }

    private func stopProgressTracking() {
        progressTimer?.invalidate()
        progressTimer = nil
    }

    // MARK: - Format Duration

    static func formatDuration(_ seconds: Double) -> String {
        let minutes = Int(seconds) / 60
        let secs = Int(seconds) % 60
        return String(format: "%d:%02d", minutes, secs)
    }

    // MARK: - Cleanup

    func clearCache() {
        for (_, url) in cachedAudioURLs {
            try? FileManager.default.removeItem(at: url)
        }
        cachedAudioURLs.removeAll()
    }

    func hasCachedAudio(for emailId: String) -> Bool {
        cachedAudioURLs[emailId] != nil
    }

    func cachedURL(for emailId: String) -> URL? {
        cachedAudioURLs[emailId]
    }
}

File: lib/services/deepgram_service.dart
import 'dart:convert';
import 'dart:io';
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:audioplayers/audioplayers.dart';
import 'package:http/http.dart' as http;

/// Deepgram Text-to-Speech Service.
/// Converts email text to audio using Deepgram's TTS API.
/// Mirrors DeepgramService.swift from the reference iOS app.
class DeepgramService extends ChangeNotifier {
  static final DeepgramService _instance = DeepgramService._internal();
  factory DeepgramService() => _instance;
  DeepgramService._internal();

  // Deepgram API key from reference project
  static const String _apiKey = '83ae019a53e7258c6fa87186d521f628dbc7eb2f';
  static const String _baseUrl = 'https://api.deepgram.com/v1/speak';

  bool _isGenerating = false;
  bool _isPlaying = false;
  double _playbackProgress = 0;
  double _audioDuration = 0;
  String? _error;

  final AudioPlayer _audioPlayer = AudioPlayer();
  Timer? _progressTimer;
  final Map<String, String> _cachedAudioPaths = {};

  // ─── Getters ─────────────────────────────────────────

  bool get isGenerating => _isGenerating;
  bool get isPlaying => _isPlaying;
  double get playbackProgress => _playbackProgress;
  double get audioDuration => _audioDuration;
  String? get error => _error;

  // ─── Generate Audio from Email ───────────────────────

  /// Generates TTS audio for an email body. Returns the local file path.
  /// Splits long text into chunks to avoid Deepgram's payload size limit.
  Future<String?> generateAudio({
    required String emailId,
    required String text,
    String voice = 'aura-asteria-en',
  }) async {
    // Return cached audio if already generated
    if (_cachedAudioPaths.containsKey(emailId)) {
      return _cachedAudioPaths[emailId];
    }

    _isGenerating = true;
    _error = null;
    notifyListeners();

    if (_apiKey.isEmpty) {
      _error = 'Deepgram API key not configured';
      _isGenerating = false;
      notifyListeners();
      return null;
    }

    try {
      // Trim to reasonable max and split into chunks
      final trimmedText = text.length > 5000 ? text.substring(0, 5000) : text;
      final chunks = _splitIntoChunks(trimmedText, maxLength: 1500);

      final allAudioData = <int>[];

      for (final chunk in chunks) {
        final chunkData = await _fetchTTSAudio(text: chunk, voice: voice);
        if (chunkData != null) {
          allAudioData.addAll(chunkData);
        }
      }

      if (allAudioData.isEmpty) {
        _error = 'Failed to generate audio';
        _isGenerating = false;
        notifyListeners();
        return null;
      }

      // Save combined audio to temp file
      final tempDir = Directory.systemTemp;
      final filePath = '${tempDir.path}/voidmail_tts_$emailId.mp3';
      final file = File(filePath);
      await file.writeAsBytes(allAudioData);
      _cachedAudioPaths[emailId] = filePath;

      _isGenerating = false;
      notifyListeners();
      return filePath;
    } catch (e) {
      debugPrint('[DeepgramService] generateAudio error: $e');
      _error = 'Failed to generate audio';
      _isGenerating = false;
      notifyListeners();
      return null;
    }
  }

  /// Fetch TTS audio for a single text chunk from Deepgram
  Future<List<int>?> _fetchTTSAudio({
    required String text,
    required String voice,
  }) async {
    try {
      final uri = Uri.parse('$_baseUrl?model=$voice&encoding=mp3');

      final response = await http.post(
        uri,
        headers: {
          'Authorization': 'Token $_apiKey',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({'text': text}),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        return response.bodyBytes;
      } else {
        debugPrint('[DeepgramService] Chunk TTS error ${response.statusCode}');
        return null;
      }
    } catch (e) {
      debugPrint('[DeepgramService] fetchTTSAudio error: $e');
      return null;
    }
  }

  /// Split text into chunks at sentence boundaries, respecting maxLength
  List<String> _splitIntoChunks(String text, {int maxLength = 1500}) {
    if (text.length <= maxLength) return [text];

    final chunks = <String>[];
    var remaining = text;

    while (remaining.isNotEmpty) {
      if (remaining.length <= maxLength) {
        chunks.add(remaining);
        break;
      }

      final searchRange = remaining.substring(0, maxLength);
      int splitIndex = -1;

      // Find the last sentence-ending punctuation within maxLength
      for (final delimiter in ['. ', '! ', '? ', '.\n', '!\n', '?\n']) {
        final idx = searchRange.lastIndexOf(delimiter);
        if (idx > splitIndex) {
          splitIndex = idx + delimiter.length;
        }
      }

      // If no sentence boundary found, split at last space
      if (splitIndex == -1) {
        final spaceIdx = searchRange.lastIndexOf(' ');
        if (spaceIdx > 0) {
          splitIndex = spaceIdx + 1;
        } else {
          // No space found — hard split
          splitIndex = maxLength;
        }
      }

      final chunk = remaining.substring(0, splitIndex).trim();
      if (chunk.isNotEmpty) {
        chunks.add(chunk);
      }
      remaining = remaining.substring(splitIndex).trim();
    }

    return chunks;
  }

  // ─── Playback Controls ───────────────────────────────

  /// Play audio from a file path
  Future<void> play(String filePath) async {
    try {
      await _audioPlayer.play(DeviceFileSource(filePath));
      _isPlaying = true;
      notifyListeners();

      // Get duration
      _audioPlayer.onDurationChanged.listen((duration) {
        _audioDuration = duration.inMilliseconds / 1000.0;
        notifyListeners();
      });

      // Start progress tracking
      _startProgressTracking();

      // Listen for completion
      _audioPlayer.onPlayerComplete.listen((_) {
        _isPlaying = false;
        _playbackProgress = 0;
        _stopProgressTracking();
        notifyListeners();
      });
    } catch (e) {
      debugPrint('[DeepgramService] play error: $e');
      _error = 'Failed to play audio';
      notifyListeners();
    }
  }

  /// Pause playback
  Future<void> pause() async {
    await _audioPlayer.pause();
    _isPlaying = false;
    _stopProgressTracking();
    notifyListeners();
  }

  /// Resume playback
  Future<void> resume() async {
    await _audioPlayer.resume();
    _isPlaying = true;
    _startProgressTracking();
    notifyListeners();
  }

  /// Stop playback
  Future<void> stop() async {
    await _audioPlayer.stop();
    _isPlaying = false;
    _playbackProgress = 0;
    _stopProgressTracking();
    notifyListeners();
  }

  /// Seek to position (0.0 to 1.0)
  Future<void> seek(double progress) async {
    final position = Duration(
      milliseconds: (progress * _audioDuration * 1000).toInt(),
    );
    await _audioPlayer.seek(position);
    _playbackProgress = progress;
    notifyListeners();
  }

  // ─── Progress Tracking ───────────────────────────────

  void _startProgressTracking() {
    _stopProgressTracking();
    _progressTimer = Timer.periodic(const Duration(milliseconds: 100), (_) async {
      final position = await _audioPlayer.getCurrentPosition();
      if (position != null && _audioDuration > 0) {
        _playbackProgress = position.inMilliseconds / (_audioDuration * 1000);
        notifyListeners();
      }
    });
  }

  void _stopProgressTracking() {
    _progressTimer?.cancel();
    _progressTimer = null;
  }

  // ─── Utilities ───────────────────────────────────────

  /// Format duration in mm:ss
  static String formatDuration(double seconds) {
    final minutes = seconds.toInt() ~/ 60;
    final secs = seconds.toInt() % 60;
    return '$minutes:${secs.toString().padLeft(2, '0')}';
  }

  /// Check if audio is cached for an email
  bool hasCachedAudio(String emailId) {
    return _cachedAudioPaths.containsKey(emailId);
  }

  /// Get cached audio path for an email
  String? cachedPath(String emailId) {
    return _cachedAudioPaths[emailId];
  }

  /// Clear all cached audio files
  Future<void> clearCache() async {
    for (final path in _cachedAudioPaths.values) {
      try {
        final file = File(path);
        if (await file.exists()) {
          await file.delete();
        }
      } catch (_) {}
    }
    _cachedAudioPaths.clear();
  }

  @override
  void dispose() {
    _audioPlayer.dispose();
    _stopProgressTracking();
    super.dispose();
  }
}
